### PR TITLE
Checkconf (#559)

### PR DESCRIPTION
**Pull request by petermax2**
Add `checkconf` interface for plugins. Enables plugin configuration validation at mount time.

See #559 .

### CODE REVIEW

File: src/libs/tools/src/backendbuilder.cpp [hunk 1]
```diff
@@ -409,6 +415,29 @@ void BackendBuilder::addPlugin (PluginSpec const & plugin)
...
+	catch (...)
+	{
+		// the checkconf operation is optional, so no worries if it was not found
+	}
+
+	if (checkConfFunction != nullptr)
+	{
+		ckdb::Key * errorKey = ckdb::keyNew (0);
+		int checkResult = checkConfFunction (errorKey, newPlugin.getConfig ().getKeySet ());
+		if (checkResult != 1)
```
Comment (markus2330) [inline src/libs/tools/src/backendbuilder.cpp]: I would do `== -1` here. What about the convention:

- 0: config was not changed (was ok)
- 1: config is changed (now ok)
- -1: config not ok, could not fix.

File: doc/tutorials/plugins.md [hunk 2]
```diff
@@ -322,6 +323,14 @@ The `elektraPluginOpen` and `elektraPluginClose` functions are not commonly used
 reviewing. `elektraPluginOpen` function runs before `elektraPluginGet` and is useful to do initialization if necessary for the plug-in. On the other
 hand `elektraPluginClose` is run after other functions of the plug-in and can be useful for freeing up resources.
 
+### elektraPluginCheckConf ###
+
+The `elektraPluginCheckConf` function may be used for validation of the plugin configuration during mount time. The signature of the function is:
+
+	int elektraLineCheckConfig (Key * errorKey, KeySet * conf)
+
+The configuration of the plugin is provided as `conf`. The function may report errors using the `errorKey` and the return value.
```
Comment (markus2330) [inline doc/tutorials/plugins.md]: Describe return values + where to read more

**Comment by markus2330** [comment]
src/plugins/doc/doc.c is missing: there should be an explanation which behaviour is expected of the function.

**Comment by petermax2** [comment]
Oops I broke something. I am going to investigate later on.

**Comment by markus2330** [comment]
Dont worry, that is what we have the build server for.

**Comment by petermax2** [comment]
jenkins build please

**Comment by petermax2** [comment]
I'm not sure how I broke it but this is the output in `CMakeErrors.log`:

	Determining if the pthread_create exist failed with the following output:
	Change Dir: /home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp

	Run Build Command:"/usr/bin/make" "cmTryCompileExec1726839152/fast"
	/usr/bin/make -f CMakeFiles/cmTryCompileExec1726839152.dir/build.make CMakeFiles/cmTryCompileExec1726839152.dir/build
	make[1]: Entering directory '/home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp'
	/usr/bin/cmake -E cmake_progress_report /home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp/CMakeFiles 1
	Building C object CMakeFiles/cmTryCompileExec1726839152.dir/CheckSymbolExists.c.o
	/usr/bin/cc   -std=gnu99  -Wno-deprecated-declarations  -Wno-long-long -pedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -fprofile-arcs -ftest-coverage -Wsign-compare -Wfloat-equal -Wformat-security    -o CMakeFiles/cmTryCompileExec1726839152.dir/CheckSymbolExists.c.o   -c /home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c
	/home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c: In function ‘main’:
	/home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:8:11: warning: ISO C forbids conversion of function pointer to object pointer type [-Wpedantic]
	   return ((int*)(&pthread_create))[argc];
		       ^
	Linking C executable cmTryCompileExec1726839152
	/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTryCompileExec1726839152.dir/link.txt --verbose=1
	/usr/bin/cc   -std=gnu99  -Wno-deprecated-declarations  -Wno-long-long -pedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -fprofile-arcs -ftest-coverage -Wsign-compare -Wfloat-equal -Wformat-security     CMakeFiles/cmTryCompileExec1726839152.dir/CheckSymbolExists.c.o  -o cmTryCompileExec1726839152 -rdynamic 
	CMakeFiles/cmTryCompileExec1726839152.dir/CheckSymbolExists.c.o: In function `main':
	CheckSymbolExists.c:(.text+0x1a): undefined reference to `pthread_create'
	collect2: error: ld returned 1 exit status
	CMakeFiles/cmTryCompileExec1726839152.dir/build.make:88: recipe for target 'cmTryCompileExec1726839152' failed
	make[1]: Leaving directory '/home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp'
	make[1]: *** [cmTryCompileExec1726839152] Error 1
	Makefile:118: recipe for target 'cmTryCompileExec1726839152/fast' failed
	make: *** [cmTryCompileExec1726839152/fast] Error 2

	File /home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:
	/* */
	#include <pthread.h>

	int main(int argc, char** argv)
	{
	  (void)argv;
	#ifndef pthread_create
	  return ((int*)(&pthread_create))[argc];
	#else
	  (void)argc;
	  return 0;
	#endif
	}

	Determining if the function pthread_create exists in the pthreads failed with the following output:
	Change Dir: /home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp

	Run Build Command:"/usr/bin/make" "cmTryCompileExec2162344864/fast"
	/usr/bin/make -f CMakeFiles/cmTryCompileExec2162344864.dir/build.make CMakeFiles/cmTryCompileExec2162344864.dir/build
	make[1]: Entering directory '/home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp'
	/usr/bin/cmake -E cmake_progress_report /home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp/CMakeFiles 1
	Building C object CMakeFiles/cmTryCompileExec2162344864.dir/CheckFunctionExists.c.o
	/usr/bin/cc   -std=gnu99  -Wno-deprecated-declarations  -Wno-long-long -pedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -fprofile-arcs -ftest-coverage -Wsign-compare -Wfloat-equal -Wformat-security -DCHECK_FUNCTION_EXISTS=pthread_create   -o CMakeFiles/cmTryCompileExec2162344864.dir/CheckFunctionExists.c.o   -c /usr/share/cmake-3.0/Modules/CheckFunctionExists.c
	Linking C executable cmTryCompileExec2162344864
	/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTryCompileExec2162344864.dir/link.txt --verbose=1
	/usr/bin/cc   -std=gnu99  -Wno-deprecated-declarations  -Wno-long-long -pedantic -Wno-variadic-macros -Wall -Wextra -Wno-overlength-strings -Wsign-compare -Wfloat-equal -Wformat-security -Wshadow -Wcomments -Wtrigraphs -Wundef -Wuninitialized -Winit-self -Wmaybe-uninitialized -fprofile-arcs -ftest-coverage -Wsign-compare -Wfloat-equal -Wformat-security -DCHECK_FUNCTION_EXISTS=pthread_create    CMakeFiles/cmTryCompileExec2162344864.dir/CheckFunctionExists.c.o  -o cmTryCompileExec2162344864 -rdynamic -lpthreads 
	/usr/bin/ld: cannot find -lpthreads
	collect2: error: ld returned 1 exit status
	CMakeFiles/cmTryCompileExec2162344864.dir/build.make:88: recipe for target 'cmTryCompileExec2162344864' failed
	make[1]: Leaving directory '/home/jenkins/workspace/workspace/elektra-mergerequests/build/CMakeFiles/CMakeTmp'
	make[1]: *** [cmTryCompileExec2162344864] Error 1
	Makefile:118: recipe for target 'cmTryCompileExec2162344864/fast' failed
	make: *** [cmTryCompileExec2162344864/fast] Error 2

Considering the changes I made, does this make any sense to you?

**Comment by petermax2** [comment]
And this is the console output:

	CMake Error at cmake/Modules/LibAddPlugin.cmake:38 (message):
	  Trying to add plugintest line but no such plugin was added (try to use
	  LINK_PLUGIN)
	Call Stack (most recent call first):
	  src/plugins/line/CMakeLists.txt:22 (add_plugintest)
	  src/plugins/CMakeLists.txt:14 (include)

It does not really correspond to CMake's error log.

**Comment by petermax2** [comment]
OK I should probably rebase ...

**Pull request closed without merging**